File: frontend/src/types/sale.ts
export type SaleOrderStatus =
  | "DRAFT"
  | "CREATED"
  | "RELEASED"
  | "IN_PROCESS"
  | "DELIVERED"
  | "INVOICED"
  | "CLOSED"
  | "CANCELLED"

export type SaleOrderType = "STANDARD" | "RUSH" | "SAMPLE" | "RETURN" | "EXPORT"
export type SaleChannel = "DIRECT" | "DISTRIBUTOR" | "ECOMMERCE" | "RETAIL"
export type SaleOrderSource = "PHONE" | "EMAIL" | "PORTAL" | "WEBSITE" | "EDI" | "MOBILE"
export type OrderPriority = "HIGH" | "MEDIUM" | "LOW"
export type SaleLineStatus =
  | "OPEN"
  | "ALLOCATED"
  | "PARTIAL"
  | "DELIVERED"
  | "INVOICED"
  | "CANCELLED"
export type CreditCheckStatus = "NOT_RUN" | "PASSED" | "FAILED" | "OVERRIDE"
export type AtpCheckStatus = "NOT_RUN" | "AVAILABLE" | "PARTIAL" | "UNAVAILABLE"
export type DocumentPaymentStatus = "UNPAID" | "PARTIAL" | "PAID"
export type DeliveryStatus = "NOT_STARTED" | "PENDING" | "PARTIAL" | "COMPLETE" | "BLOCKED"
export type BillingStatus = "NOT_INVOICED" | "PARTIAL" | "INVOICED" | "COMPLETE"
export type SalePartnerRole =
  | "SOLD_TO"
  | "SHIP_TO"
  | "BILL_TO"
  | "PAYER"
  | "FORWARDING_AGENT"
  | "SALES_EMPLOYEE"

export type AddressBlock = {
  line1?: string | null
  line2?: string | null
  city?: string | null
  state?: string | null
  postal_code?: string | null
  country?: string | null
}

export type CustomerSnapshot = {
  customer_id?: number | null
  customer_code?: string | null
  name?: string | null
  billing_address?: AddressBlock | null
  shipping_address?: AddressBlock | null
  contact_name?: string | null
  contact_phone?: string | null
  contact_email?: string | null
  customer_group?: string | null
  credit_limit?: string | null
  credit_status?: string | null
  payment_terms?: string | null
  tax_id?: string | null
  gst_number?: string | null
  vat_number?: string | null
  shipping_preference?: string | null
  incoterms?: string | null
}

export type SalePartnerInput = {
  role: SalePartnerRole
  customer_id?: number | null
  supplier_id?: number | null
  user_id?: number | null
  name_override?: string | null
  address?: AddressBlock | null
}

export type SalePartner = SalePartnerInput & {
  id: number
  customer_name?: string | null
  supplier_name?: string | null
  user_email?: string | null
}

export type SaleOrderSummary = {
  total_items: number
  total_quantity: number
  total_net: string
  total_tax: string
  total_discount: string
  freight: string
  insurance: string
  handling: string
  grand_total: string
}

export type SaleItemLine = {
  id?: number
  product_id: number
  quantity: number
  unit_price?: string | null
  description?: string | null
  uom?: string | null
  alternate_uom?: string | null
  uom_conversion_factor?: string | null
  discount_percent?: string
  discount_amount?: string
  tax_code?: string | null
  tax_rate_id?: number | null
  requested_delivery_date?: string | null
  confirmed_delivery_date?: string | null
  product_category?: string | null
  item_category?: string | null
  gross_price?: string | null
  warehouse_id?: number | null
  storage_location_id?: number | null
  batch_number?: string | null
  serial_number?: string | null
  delivery_block?: string | null
  billing_block?: string | null
  rejection_reason?: string | null
  net_weight?: string | null
  gross_weight?: string | null
  volume?: string | null
  substitute_product_id?: number | null
  line_text?: string | null
  line_status?: SaleLineStatus | null
}

export type SaleItem = {
  id: number
  line_number: number
  product_id: number
  product_name: string
  product_sku: string
  description: string | null
  quantity: number
  uom: string | null
  alternate_uom: string | null
  uom_conversion_factor: string | null
  unit_price: string
  price_at_sale: string
  gross_price: string | null
  discount_percent: string
  discount_amount: string
  tax_code: string | null
  tax_rate_id: number | null
  tax_amount: string
  net_amount: string
  line_total: string
  requested_delivery_date: string | null
  confirmed_delivery_date: string | null
  confirmed_quantity: number
  delivered_quantity: number
  line_status: SaleLineStatus
  backorder_quantity: number
  product_category: string | null
  item_category: string | null
  warehouse_id: number | null
  warehouse_name: string | null
  storage_location_id: number | null
  batch_number: string | null
  serial_number: string | null
  delivery_block: string | null
  billing_block: string | null
  rejection_reason: string | null
  net_weight: string | null
  gross_weight: string | null
  volume: string | null
  substitute_product_id: number | null
  substitute_product_sku: string | null
  line_text: string | null
}

export type SaleOrder = {
  id: number
  order_number: string
  order_status: SaleOrderStatus
  order_date: string
  order_type: SaleOrderType
  sales_channel: SaleChannel | null
  order_source: SaleOrderSource | null
  priority: OrderPriority
  salesperson_id: number | null
  salesperson_email: string | null
  is_pos_checkout: boolean
  customer_id: number | null
  customer_name: string | null
  cashier_email: string | null
  created_at: string
  updated_at: string | null
  updated_by_email: string | null
  items: SaleItem[]
  partners: SalePartner[]
  summary: SaleOrderSummary | null
  gross_total: string
  header_discount_amount: string
  freight_amount: string
  insurance_amount: string
  handling_amount: string
  subtotal: string
  tax: string
  total: string
  amount_paid: string
  payment_status: DocumentPaymentStatus
  currency_code: string
  price_list_code: string | null
  pricing_procedure: string | null
  payment_terms: string | null
  payment_due_date: string | null
  payment_method_id: number | null
  warehouse_id: number | null
  warehouse_name: string | null
  shipping_point_id: number | null
  shipping_point_name: string | null
  partial_delivery_allowed: boolean
  complete_delivery_required: boolean
  planned_ship_date: string | null
  requested_delivery_date: string | null
  shipping_method: string | null
  shipping_conditions: string | null
  transportation_group: string | null
  loading_group: string | null
  incoterms: string | null
  incoterms_location: string | null
  delivery_block: string | null
  sales_organization: string | null
  distribution_channel: string | null
  division: string | null
  sales_office: string | null
  sales_group: string | null
  customer_po_number: string | null
  customer_po_date: string | null
  opportunity_id: number | null
  campaign_id: string | null
  price_group: string | null
  header_text: string | null
  credit_check_status: CreditCheckStatus
  atp_check_status: AtpCheckStatus
  invoice_status: BillingStatus | null
  delivery_status: DeliveryStatus | null
  approval_status: string | null
  customer_snapshot: Record<string, unknown> | null
  pricing_conditions: Record<string, unknown> | null
  delivery_logistics: Record<string, unknown> | null
  billing_financial: Record<string, unknown> | null
  terms_compliance: Record<string, unknown> | null
  references: Record<string, unknown> | null
  attachments: Record<string, unknown>[] | null
  workflow_history: Record<string, unknown>[] | null
}

export type SaleListItem = {
  id: number
  order_number: string
  order_status: SaleOrderStatus
  order_type: SaleOrderType
  customer_id: number | null
  customer_name: string | null
  cashier_email: string | null
  created_at: string
  order_date: string
  item_count: number
  total: string
  payment_status: DocumentPaymentStatus
  currency_code: string
  delivery_status: DeliveryStatus | null
}

export type SaleListResponse = {
  items: SaleListItem[]
  total: number
}

export type SaleLookups = {
  customers: { id: number; customer_code: string | null; name: string }[]
  warehouses: { id: number; code: string; name: string }[]
  users: { id: number; email: string }[]
  tax_rates: { id: number; code: string; name: string; rate_percent: string }[]
  payment_methods: { id: number; code: string; name: string }[]
}

export type SaleOrderPayload = {
  order_number_override?: string | null
  customer_id?: number | null
  order_date?: string | null
  order_type?: SaleOrderType
  sales_channel?: SaleChannel | null
  order_source?: SaleOrderSource | null
  priority?: OrderPriority
  salesperson_id?: number | null
  currency_code?: string | null
  price_list_code?: string | null
  pricing_procedure?: string | null
  payment_terms?: string | null
  payment_due_date?: string | null
  payment_method_id?: number | null
  header_discount_amount?: string
  freight_amount?: string
  insurance_amount?: string
  handling_amount?: string
  warehouse_id?: number | null
  shipping_point_id?: number | null
  partial_delivery_allowed?: boolean
  complete_delivery_required?: boolean
  planned_ship_date?: string | null
  requested_delivery_date?: string | null
  shipping_method?: string | null
  shipping_conditions?: string | null
  transportation_group?: string | null
  loading_group?: string | null
  incoterms?: string | null
  incoterms_location?: string | null
  delivery_block?: string | null
  sales_organization?: string | null
  distribution_channel?: string | null
  division?: string | null
  sales_office?: string | null
  sales_group?: string | null
  customer_po_number?: string | null
  customer_po_date?: string | null
  opportunity_id?: number | null
  campaign_id?: string | null
  price_group?: string | null
  header_text?: string | null
  approval_status?: string | null
  customer_snapshot?: CustomerSnapshot | null
  pricing_conditions?: Record<string, unknown> | null
  delivery_logistics?: Record<string, unknown> | null
  billing_financial?: Record<string, unknown> | null
  terms_compliance?: Record<string, unknown> | null
  references?: Record<string, unknown> | null
  workflow_approval?: Record<string, unknown> | null
  attachments?: { filename: string; url?: string | null; content_type?: string | null }[] | null
  partners?: SalePartnerInput[] | null
  items: SaleItemLine[]
  confirm?: boolean
}

export type SaleCheckoutPayload = {
  customer_id?: number | null
  items: { product_id: number; quantity: number }[]
  confirm?: boolean
}

export type PosProduct = {
  id: number
  sku: string
  name: string
  barcode: string | null
  price: string
  stock: number
  low_stock_threshold: number
}

export type PosProductListResponse = {
  items: PosProduct[]
  total: number
}

/** @deprecated Use SaleOrder */
export type Sale = SaleOrder

File: backend/app/schemas/sale.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import (
    AtpCheckStatus,
    BillingStatus,
    CreditCheckStatus,
    DeliveryStatus,
    DocumentPaymentStatus,
    OrderPriority,
    SaleChannel,
    SaleLineStatus,
    SaleOrderSource,
    SaleOrderStatus,
    SaleOrderType,
    SalePartnerRole,
)


class AddressBlock(BaseModel):
    line1: str | None = None
    line2: str | None = None
    city: str | None = None
    state: str | None = None
    postal_code: str | None = None
    country: str | None = None


class CustomerSnapshot(BaseModel):
    customer_id: int | None = None
    customer_code: str | None = None
    name: str | None = None
    billing_address: AddressBlock | None = None
    shipping_address: AddressBlock | None = None
    contact_name: str | None = None
    contact_phone: str | None = None
    contact_email: str | None = None
    customer_group: str | None = None
    credit_limit: Decimal | None = None
    credit_status: str | None = None
    payment_terms: str | None = None
    tax_id: str | None = None
    shipping_preference: str | None = None
    incoterms: str | None = None
    trade_name: str | None = None
    status: str | None = None
    gst_number: str | None = None
    vat_number: str | None = None
    currency_code: str | None = None


class PricingConditions(BaseModel):
    contract_ref: str | None = None
    special_pricing_agreement: str | None = None
    total_before_tax: Decimal | None = None
    total_after_tax: Decimal | None = None
    line_discount_notes: str | None = None
    pricing_date: date | None = None
    manual_price_override_reason: str | None = None
    list_price_notes: str | None = None


class DeliveryLogistics(BaseModel):
    ship_to_override: AddressBlock | None = None
    shipping_instructions: str | None = None
    route: str | None = None
    warehouse_assignment: str | None = None
    promised_date: date | None = None
    actual_date: date | None = None
    delivery_number: str | None = None
    carrier_name: str | None = None
    tracking_number: str | None = None
    transport_mode: str | None = None
    bill_of_lading: str | None = None
    packaging_details: str | None = None
    expected_shipment_date: date | None = None


class BillingFinancial(BaseModel):
    bill_to: AddressBlock | None = None
    invoice_type: str | None = None
    cost_center: str | None = None
    profit_center: str | None = None
    project_code: str | None = None
    wbs_element: str | None = None
    invoice_number: str | None = None
    invoice_date: date | None = None
    billing_type: str | None = None
    taxable_amount: Decimal | None = None
    tax_amount: Decimal | None = None
    total_amount: Decimal | None = None
    accounting_document_number: str | None = None
    payment_method_label: str | None = None
    payment_receipt_number: str | None = None
    amount_received: Decimal | None = None
    outstanding_amount: Decimal | None = None
    credit_debit_note_ref: str | None = None
    billing_block: str | None = None
    split_billing: bool | None = None
    bank_details: str | None = None
    card_last_four: str | None = None


class TermsCompliance(BaseModel):
    warranty_terms: str | None = None
    return_policy_ref: str | None = None
    inspection_requirements: str | None = None
    regulatory_compliance: str | None = None
    order_validity_date: date | None = None
    order_expiry_date: date | None = None
    export_license: str | None = None
    hazardous_material_info: str | None = None
    sustainability_data: str | None = None


class SaleReferences(BaseModel):
    customer_po_number: str | None = None
    customer_po_date: date | None = None
    quotation_number: str | None = None
    contract_number: str | None = None
    tender_number: str | None = None
    internal_order_number: str | None = None
    remarks: str | None = None
    internal_notes: str | None = None
    reason_code: str | None = None
    order_reason: str | None = None
    attachments_note: str | None = None
    atp_check_notes: str | None = None
    crm_opportunity_link: str | None = None
    forecast_vs_actual: str | None = None
    forecast_category: str | None = None
    closing_probability: str | None = None
    expected_revenue: str | None = None
    lead_source: str | None = None
    order_confirmation_date: date | None = None
    returns_cancellation_reason: str | None = None
    fulfillment_lead_time_days: str | None = None
    profitability_analysis: str | None = None


class AttachmentMeta(BaseModel):
    filename: str
    url: str | None = None
    content_type: str | None = None


class WorkflowEvent(BaseModel):
    at: datetime | None = None
    status: str
    user_email: str | None = None
    note: str | None = None
    approver_name: str | None = None
    approver_date: date | None = None
    credit_approval_status: str | None = None
    margin_approval_status: str | None = None
    legal_approval_status: str | None = None


class WorkflowApproval(BaseModel):
    approver_name: str | None = None
    approver_date: date | None = None
    credit_approval_status: str | None = None
    margin_approval_status: str | None = None
    legal_approval_status: str | None = None
    note: str | None = None


class SalePartnerCreate(BaseModel):
    role: SalePartnerRole
    customer_id: int | None = None
    supplier_id: int | None = None
    user_id: int | None = None
    name_override: str | None = None
    address: AddressBlock | None = None


class SalePartnerRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    role: SalePartnerRole
    customer_id: int | None = None
    customer_name: str | None = None
    supplier_id: int | None = None
    supplier_name: str | None = None
    user_id: int | None = None
    user_email: str | None = None
    name_override: str | None = None
    address: dict[str, Any] | None = None


class SaleItemLineCreate(BaseModel):
    product_id: int = Field(ge=1)
    quantity: int = Field(ge=1)
    unit_price: Decimal | None = None
    description: str | None = None
    uom: str | None = None
    alternate_uom: str | None = None
    uom_conversion_factor: Decimal | None = None
    discount_percent: Decimal = Field(default=Decimal("0"), ge=0, le=100)
    discount_amount: Decimal = Field(default=Decimal("0"), ge=0)
    tax_code: str | None = None
    tax_rate_id: int | None = None
    requested_delivery_date: date | None = None
    confirmed_delivery_date: date | None = None
    product_category: str | None = None
    item_category: str | None = None
    gross_price: Decimal | None = None
    warehouse_id: int | None = None
    storage_location_id: int | None = None
    batch_number: str | None = None
    serial_number: str | None = None
    delivery_block: str | None = None
    billing_block: str | None = None
    rejection_reason: str | None = None
    net_weight: Decimal | None = None
    gross_weight: Decimal | None = None
    volume: Decimal | None = None
    substitute_product_id: int | None = None
    line_text: str | None = None
    line_status: SaleLineStatus | None = None


class SaleItemLineUpdate(SaleItemLineCreate):
    id: int | None = None


def _validate_b2b_fields(
    *,
    customer_id: int | None,
    customer_po_number: str | None,
    payment_terms: str | None,
    sales_organization: str | None,
    items: list[SaleItemLineCreate],
) -> None:
    missing: list[str] = []
    if not customer_id:
        missing.append("customer_id")
    if not (customer_po_number or "").strip():
        missing.append("customer_po_number")
    if not (payment_terms or "").strip():
        missing.append("payment_terms")
    if not (sales_organization or "").strip():
        missing.append("sales_organization")
    if not items:
        missing.append("items")
    else:
        for idx, line in enumerate(items):
            if line.product_id < 1:
                missing.append(f"items[{idx}].product_id")
            if line.quantity < 1:
                missing.append(f"items[{idx}].quantity")
            if line.unit_price is None:
                missing.append(f"items[{idx}].unit_price")
    if missing:
        raise ValueError(f"B2B order missing required fields: {', '.join(missing)}")


class SaleOrderBase(BaseModel):
    customer_id: int | None = None
    order_date: date | None = None
    order_type: SaleOrderType = SaleOrderType.STANDARD
    sales_channel: SaleChannel | None = None
    order_source: SaleOrderSource | None = None
    priority: OrderPriority = OrderPriority.MEDIUM
    salesperson_id: int | None = None
    currency_code: str | None = None
    price_list_code: str | None = None
    pricing_procedure: str | None = None
    payment_terms: str | None = None
    payment_due_date: date | None = None
    payment_method_id: int | None = None
    header_discount_amount: Decimal = Field(default=Decimal("0"), ge=0)
    freight_amount: Decimal = Field(default=Decimal("0"), ge=0)
    insurance_amount: Decimal = Field(default=Decimal("0"), ge=0)
    handling_amount: Decimal = Field(default=Decimal("0"), ge=0)
    warehouse_id: int | None = None
    shipping_point_id: int | None = None
    partial_delivery_allowed: bool = True
    complete_delivery_required: bool = False
    planned_ship_date: date | None = None
    requested_delivery_date: date | None = None
    shipping_method: str | None = None
    shipping_conditions: str | None = None
    transportation_group: str | None = None
    loading_group: str | None = None
    incoterms: str | None = None
    incoterms_location: str | None = None
    delivery_block: str | None = None
    sales_organization: str | None = None
    distribution_channel: str | None = None
    division: str | None = None
    sales_office: str | None = None
    sales_group: str | None = None
    customer_po_number: str | None = None
    customer_po_date: date | None = None
    opportunity_id: int | None = None
    campaign_id: str | None = None
    price_group: str | None = None
    header_text: str | None = None
    approval_status: str | None = None
    customer_snapshot: CustomerSnapshot | None = None
    pricing_conditions: PricingConditions | None = None
    delivery_logistics: DeliveryLogistics | None = None
    billing_financial: BillingFinancial | None = None
    terms_compliance: TermsCompliance | None = None
    references: SaleReferences | None = None
    attachments: list[AttachmentMeta] | None = None
    partners: list[SalePartnerCreate] | None = None
    workflow_approval: WorkflowApproval | None = None
    require_b2b_fields: bool = True


class SaleOrderCreate(SaleOrderBase):
    items: list[SaleItemLineCreate] = Field(min_length=1)
    confirm: bool = False
    order_number_override: str | None = Field(default=None, max_length=32)

    @model_validator(mode="after")
    def validate_b2b(self) -> SaleOrderCreate:
        if not self.require_b2b_fields:
            return self
        _validate_b2b_fields(
            customer_id=self.customer_id,
            customer_po_number=self.customer_po_number,
            payment_terms=self.payment_terms,
            sales_organization=self.sales_organization,
            items=self.items,
        )
        return self


class SaleOrderUpdate(SaleOrderBase):
    items: list[SaleItemLineUpdate] | None = None
    require_b2b_fields: bool = True


class SaleConfirmRequest(BaseModel):
    run_credit_check: bool = True
    run_atp_check: bool = True
    override_credit_failure: bool = False


class PosProductRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    sku: str
    name: str
    barcode: str | None
    price: Decimal
    stock: int
    low_stock_threshold: int


class PosProductListResponse(BaseModel):
    items: list[PosProductRead]
    total: int


class PosCheckoutCreate(BaseModel):
    customer_id: int | None = None
    items: list[SaleItemLineCreate] = Field(min_length=1)
    confirm: bool = True


class SaleOrderSummary(BaseModel):
    total_items: int
    total_quantity: int
    total_net: Decimal
    total_tax: Decimal
    total_discount: Decimal
    freight: Decimal
    insurance: Decimal
    handling: Decimal
    grand_total: Decimal


class SaleItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    line_number: int
    product_id: int
    product_name: str
    product_sku: str
    description: str | None = None
    quantity: int
    uom: str | None = None
    alternate_uom: str | None = None
    uom_conversion_factor: Decimal | None = None
    unit_price: Decimal
    price_at_sale: Decimal
    gross_price: Decimal | None = None
    discount_percent: Decimal
    discount_amount: Decimal
    tax_code: str | None = None
    tax_rate_id: int | None = None
    tax_amount: Decimal
    net_amount: Decimal
    line_total: Decimal
    requested_delivery_date: date | None = None
    confirmed_delivery_date: date | None = None
    confirmed_quantity: int = 0
    delivered_quantity: int = 0
    line_status: SaleLineStatus
    backorder_quantity: int
    product_category: str | None = None
    item_category: str | None = None
    warehouse_id: int | None = None
    warehouse_name: str | None = None
    storage_location_id: int | None = None
    batch_number: str | None = None
    serial_number: str | None = None
    delivery_block: str | None = None
    billing_block: str | None = None
    rejection_reason: str | None = None
    net_weight: Decimal | None = None
    gross_weight: Decimal | None = None
    volume: Decimal | None = None
    substitute_product_id: int | None = None
    substitute_product_sku: str | None = None
    line_text: str | None = None


class SaleOrderRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    order_number: str
    order_status: SaleOrderStatus
    order_date: date
    order_type: SaleOrderType
    sales_channel: SaleChannel | None
    order_source: SaleOrderSource | None
    priority: OrderPriority
    salesperson_id: int | None = None
    salesperson_email: str | None = None
    is_pos_checkout: bool
    customer_id: int | None
    customer_name: str | None = None
    cashier_email: str | None = None
    created_at: datetime
    updated_at: datetime | None = None
    updated_by_email: str | None = None
    items: list[SaleItemRead]
    partners: list[SalePartnerRead] = []
    summary: SaleOrderSummary | None = None
    gross_total: Decimal
    header_discount_amount: Decimal
    freight_amount: Decimal
    insurance_amount: Decimal
    handling_amount: Decimal
    subtotal: Decimal
    tax: Decimal
    total: Decimal
    amount_paid: Decimal
    payment_status: DocumentPaymentStatus
    currency_code: str
    price_list_code: str | None = None
    pricing_procedure: str | None = None
    payment_terms: str | None = None
    payment_due_date: date | None = None
    payment_method_id: int | None = None
    warehouse_id: int | None = None
    warehouse_name: str | None = None
    shipping_point_id: int | None = None
    shipping_point_name: str | None = None
    partial_delivery_allowed: bool
    complete_delivery_required: bool = False
    planned_ship_date: date | None = None
    requested_delivery_date: date | None = None
    shipping_method: str | None = None
    shipping_conditions: str | None = None
    transportation_group: str | None = None
    loading_group: str | None = None
    incoterms: str | None = None
    incoterms_location: str | None = None
    delivery_block: str | None = None
    sales_organization: str | None = None
    distribution_channel: str | None = None
    division: str | None = None
    sales_office: str | None = None
    sales_group: str | None = None
    customer_po_number: str | None = None
    customer_po_date: date | None = None
    opportunity_id: int | None = None
    campaign_id: str | None = None
    price_group: str | None = None
    header_text: str | None = None
    credit_check_status: CreditCheckStatus
    atp_check_status: AtpCheckStatus
    invoice_status: BillingStatus | None = None
    delivery_status: DeliveryStatus | None = None
    approval_status: str | None = None
    customer_snapshot: dict[str, Any] | None = None
    pricing_conditions: dict[str, Any] | None = None
    delivery_logistics: dict[str, Any] | None = None
    billing_financial: dict[str, Any] | None = None
    terms_compliance: dict[str, Any] | None = None
    references: dict[str, Any] | None = None
    attachments: list[dict[str, Any]] | None = None
    workflow_history: list[dict[str, Any]] | None = None


class SaleListItem(BaseModel):
    id: int
    order_number: str
    order_status: SaleOrderStatus
    order_type: SaleOrderType
    customer_id: int | None
    customer_name: str | None
    cashier_email: str | None
    created_at: datetime
    order_date: date
    item_count: int
    total: Decimal
    payment_status: DocumentPaymentStatus
    currency_code: str
    delivery_status: DeliveryStatus | None = None


class SaleListResponse(BaseModel):
    items: list[SaleListItem]
    total: int


class SaleLookupCustomer(BaseModel):
    id: int
    customer_code: str | None
    name: str


class SaleLookupWarehouse(BaseModel):
    id: int
    code: str
    name: str


class SaleLookupUser(BaseModel):
    id: int
    email: str


class SaleLookupTaxRate(BaseModel):
    id: int
    code: str
    name: str
    rate_percent: Decimal


class SaleLookupPaymentMethod(BaseModel):
    id: int
    code: str
    name: str


class SaleLookupsResponse(BaseModel):
    customers: list[SaleLookupCustomer]
    warehouses: list[SaleLookupWarehouse]
    users: list[SaleLookupUser]
    tax_rates: list[SaleLookupTaxRate]
    payment_methods: list[SaleLookupPaymentMethod] = []
